### PR TITLE
Added animations to player and mech boss

### PR DESCRIPTION
**Pull request by MitchellCoon**

### CODE REVIEW

File: Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller [hunk 1]
```diff
@@ -0,0 +1,1045 @@
+%YAML 1.1
```
Comment (townofdon) [inline Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller:1]: Totally fine for this gamejam - instead of multiple animators, using an [AnimatorOverrideController](https://docs.unity3d.com/Manual/AnimatorOverrideController.html) or a [BlendTree](https://docs.unity3d.com/Manual/class-BlendTree.html) would provide a more scalable / modular way of doing this in the future.

File: Assets/Scripts/Bosses/MechBossMissileMotion.cs [hunk 2]
```diff
@@ -30,7 +31,7 @@ void Update()
     {
         if (!isFalling && Time.time - spawnTime >= riseTime)
         {
-            transform.position = new Vector3(playerLocationAtSpawn.x, fallHeight, transform.position.z);
+            transform.position = new Vector3(playerLocationAtSpawn.x + horizontalOffset, fallHeight, transform.position.z);
             Vector3 localScale = transform.localScale;
 		    localScale.y *= -1;
 		    transform.localScale = localScale;
```
Comment (townofdon) [inline Assets/Scripts/Bosses/MechBossMissileMotion.cs:37]: Since we're managing the `transform.position` directly, we'll want to make sure that the missile's RigidBody2D is set to `kinematic` if not already so

File: Assets/Scripts/Bosses/MissileTelegraph.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+using UnityEngine;
+
+public class MissileTelegraph : MonoBehaviour
+{
+    [SerializeField] float duration;
+    float spawnTime;
+
+    void Start()
+    {
+        spawnTime = Time.time;
```
Comment (townofdon) [inline Assets/Scripts/Bosses/MissileTelegraph.cs:12]: This is extremely minor, but this method is subject to floating point precision loss.

Starting at a value of `0` would be a way to avoid this. Probably doesn't matter for this gamejam though.

Comment (townofdon) [reply Assets/Scripts/Bosses/MissileTelegraph.cs:12]: e.g.

```
float timeElapsed = 0f;

void Update() {
  timeElapsed += Time.deltaTime;
  if (timeElapsed > duration) Splode();
}
```

File: Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs [hunk 4]
```diff
@@ -21,4 +22,9 @@ void Update()
             Destroy(gameObject);
         }
     }
+
+    public void SetPossessionManager(PossessionManager sourcePossessionManager)
+    {
+        possessionManager = sourcePossessionManager;
+    }
```
Comment (townofdon) [inline Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs:29]: Why is this needed? Is this for the projectile? It seems weird to me that the projectile should need to know any information from the possession manager.

File: Assets/Scripts/Capabilities/Jump.cs [hunk 5]
```diff
@@ -2,11 +2,15 @@
 using System.Collections.Generic;
 using UnityEngine;
 
+public enum JumpState {Rising, Falling, Grounded};
```
Comment (townofdon) [inline Assets/Scripts/Capabilities/Jump.cs:5]: nit (minor): the first enum state will be the default value - I think that Grounded would be the default here. Probably totally fine

File: Assets/Scripts/Capabilities/Jump.cs [hunk 6]
```diff
@@ -64,6 +69,12 @@ private void FixedUpdate()
         if (isGrounded)
         {
             jumpPhase = 0;
+            if (jumpState != JumpState.Grounded)
+            {
+                animator.SetBool(Constants.JUMP_FALL_ANIMATION, false);
+                animator.SetBool(Constants.JUMP_LAND_ANIMATION, true);
+            }
+            jumpState = JumpState.Grounded;
```
Comment (townofdon) [inline Assets/Scripts/Capabilities/Jump.cs:77]: Having two variables keep track of the grounded state (`isGrounded`, `jumpState.Grounded`) might introduce some bugs if we're not careful. Could we replace `isGrounded` with this new jumpState? No worries if that would be too much effort.

File: Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller (hunk 1 above)
Comment (MitchellCoon) [reply Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller:1]: I was not aware of this feature. I'll make a note of it for the future but probably not gonna update it before the deadline

File: Assets/Scripts/Bosses/MechBossMissileMotion.cs (hunk 2 above)
Comment (MitchellCoon) [reply Assets/Scripts/Bosses/MechBossMissileMotion.cs:37]: What exactly does that mean?

File: Assets/Scripts/Capabilities/Move.cs [hunk 7]
```diff
@@ -43,6 +45,9 @@ private void FixedUpdate()
 
         body.velocity = velocity;
 
+        isMoving = Mathf.Abs(velocity.x) > 0 ? true : false;
```
Comment (townofdon) [inline Assets/Scripts/Capabilities/Move.cs:48]: nit (minor): I've seen tutorials recommend using [Mathf.Epsilon](https://docs.unity3d.com/ScriptReference/Mathf.Epsilon.html) instead of `0`, because sometimes a value can be verrrry close to zero but not zero.

File: Assets/Scripts/Bosses/MissileTelegraph.cs (hunk 3 above)
Comment (MitchellCoon) [reply Assets/Scripts/Bosses/MissileTelegraph.cs:12]: Not sure if I understand the suggestion. How would we keep track of the time since spawning if we set the spawnTime to zero?

File: Assets/Scripts/Combat/Hitbox.cs [hunk 8]
```diff
@@ -10,14 +10,17 @@ public class Hitbox : DisableSpriteRender
     public bool isEnemyHitbox = false;
     public bool hitOnce = false;
 
+    [SerializeField] PossessionManager possessionManager;
```
Comment (townofdon) [inline Assets/Scripts/Combat/Hitbox.cs:13]: Nit (minor): this variable is named the same as the [out variable on line 27](https://github.com/MitchellCoon/MetroidvaniaGameJamFeb23/pull/45/files#diff-e46262b23de53ed8582a70b8be4c5a060b9595f97b168768c795e3386ecb9b22L24) and introduces variable shadowing - could we name this something different? Even `_posessionManager` with an underscore would be enough.

File: Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs (hunk 4 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs:29]: This was part of the workaround for the hitbox logic. The projectile needs to know if the enemy that spawned it was possessed or not so it can determine what should be taking damage when it hits something

File: Assets/Scripts/Combat/Hitbox.cs [hunk 9]
```diff
@@ -10,14 +10,17 @@ public class Hitbox : DisableSpriteRender
...
     {
         attackData = newAttackData;
     }
 
     void OnTriggerEnter2D(Collider2D other)
     {
-        if (!isEnemyHitbox && other.CompareTag(Constants.ENEMY_TAG))
+        if ((!isEnemyHitbox && other.CompareTag(Constants.ENEMY_TAG)) || (isEnemyHitbox && possessionManager != null && possessionManager.IsPossessed()))
         {
             if (attackData.willPossessTarget && other.TryGetComponent<PossessionManager>(out var possessionManager))
```
Comment (townofdon) [inline Assets/Scripts/Combat/Hitbox.cs:25]: Alternatively we could rename this out var

File: Assets/Scripts/Capabilities/Jump.cs (hunk 5 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/Jump.cs:5]: I just put them in the order that the animation follows, but that makes sense

File: Assets/Scripts/Capabilities/Jump.cs (hunk 6 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/Jump.cs:77]: I can probably do that

**Review by townofdon — APPROVED** [review]
Looks good! Added a bunch of comments, mostly minor I think

File: Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller (hunk 1 above)
Comment (townofdon) [reply Assets/Animations/Characters/Slime/Slime Hurt Animations/SlimeHurtAnimator.controller:1]: Oh absolutely, don't change anything in the current implementation if it already works.

File: Assets/Scripts/Capabilities/Move.cs (hunk 7 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/Move.cs:48]: That makes sense, I'll update that

File: Assets/Scripts/Bosses/MechBossMissileMotion.cs (hunk 2 above)
Comment (townofdon) [reply Assets/Scripts/Bosses/MechBossMissileMotion.cs:37]: From the [docs](https://docs.unity3d.com/Manual/RigidbodiesOverview.html):

> In most cases, if a GameObject has a Rigidbody, you should use the Rigidbody properties to move the GameObject, instead of the Transform properties. The Rigidbody properties apply forces and torque from the physics system, which change the GameObject’s Transform; if you then change the Transform directly, Unity cannot correctly calculate the physics simulation, and you might see unwanted behavior. This is particularly true of rigid bodies with [Joints](https://docs.unity3d.com/Manual/joints-section.html).

> There are some cases where you might want the physics system to detect a GameObject, but not to control it. For example, you might want [Colliders](https://docs.unity3d.com/Manual/collision-section.html)
 to detect a GameObject, but you intend to control that GameObject’s movement and position via its [Transform](https://docs.unity3d.com/Manual/class-Transform.html).

> Movement in Unity that is not physics-based is called kinematic motion. The Rigidbody component has the property Is Kinematic which, when enabled, defines the attached GameObject as non-physics-based, and removes it from the control of the physics engine. This allows you to move it kinematically via the Transform without Unity’s physics simulation calculations overriding the changes.

File: Assets/Scripts/Bosses/MissileTelegraph.cs (hunk 3 above)
Comment (townofdon) [reply Assets/Scripts/Bosses/MissileTelegraph.cs:12]: `timeElapsed` would be the time since spawning. It will initialize to `0` and increment by `Time.deltaTime` each frame. And then continue incrementing until it is equal to or greater than `duration`.

File: Assets/Scripts/Combat/Hitbox.cs (hunk 8 above)
Comment (MitchellCoon) [reply Assets/Scripts/Combat/Hitbox.cs:13]: I think it would make sense to have one as the sourcePossessionManager and the other as the targetPossessionManager

File: Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs (hunk 4 above)
Comment (townofdon) [reply Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs:29]: Okay, this works then. Thanks for explaining.

As an aside, one way I've circumvented this issue in the past was by registering a GUID on the parent object, and set the same GUID on the projectile. Then, in the `OnTriggerEnter` (or equivalent) check, if the guids are equal, ignore the collision.

File: Assets/Scripts/Capabilities/Jump.cs (hunk 5 above)
Comment (townofdon) [reply Assets/Scripts/Capabilities/Jump.cs:5]: No change needed. In fact, highly recommend NOT changing enum values.

File: Assets/Scripts/Player/PlayerCombat.cs [hunk 10]
```diff
@@ -16,6 +19,8 @@ public class PlayerCombat : MonoBehaviour
     PlayerMain playerMain;
 
     bool isAlive;
+    int hurtThreshold = 7;
+    int perilThreshold = 3;
```
[the remark this answers is not in the document]
Comment (MitchellCoon) [reply Assets/Scripts/Player/PlayerCombat.cs:23]: Good call, must've been tired when I wrote that

File: Assets/Scripts/Bosses/MissileTelegraph.cs (hunk 3 above)
Comment (MitchellCoon) [reply Assets/Scripts/Bosses/MissileTelegraph.cs:12]: Ignore my last comment, for some reason your second comment wasn't displaying when I wrote that. This makes sense

File: Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs (hunk 4 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/BaseEnemyProjectileMotion.cs:29]: I'm not sure if that would solve everything. It would prevent enemies from killing themselves with their own projectiles, but we would still need a way to ensure that possessed enemy projectiles damage other enemies. Possession is a lot more complicated than I thought it would be

File: Assets/Scripts/Capabilities/Jump.cs (hunk 5 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/Jump.cs:5]: For now these one are used exclusively within the Jump script, so if ever there was a time to reorder them, it would be now

Comment (townofdon) [reply Assets/Scripts/Capabilities/Jump.cs]: Sounds good, no reorder needed. Since enums are just basically integers under the hood, changing enum order will mean that, for example

```
// numbers just added as an example

enum BfastFood {
  Saugage = 0,
  Eggs = 1,
  Bacon = 2,
}

// changed to:

enum BfastFood {
  Pancakes = 0, // << added
  Saugage = 1,
  Eggs = 2,
  Bacon = 3,
}

// let's say we previously serialized this, and it was set to Bacon (2)
[SerializeField] BfastFood food;

// after changing the enum, the value would now be Eggs (2)
```

Hopefully that makes sense.

**Review by townofdon — APPROVED** [review]

**Comment by townofdon** [comment]
:shipit:

File: Assets/Scripts/Capabilities/Jump.cs (hunk 5 above)
Comment (MitchellCoon) [reply Assets/Scripts/Capabilities/Jump.cs]: That does make sense. Good to know for the future
